File: machine/cpp/machine/devices/audio/output_ring.cpp
#include "machine/devices/audio/output_ring.h"

#include <algorithm>

namespace bmsx {

void ApuOutputRing::clear() {
	m_readFrame = 0;
	m_queuedFrames = 0;
	m_firstFrameSequence = 0;
}

size_t ApuOutputRing::queuedFrames() const {
	return m_queuedFrames;
}

i64 ApuOutputRing::firstFrameSequence() const {
	return m_firstFrameSequence;
}

void ApuOutputRing::write(const i16* samples, size_t frameCount, i64 startSequence) {
	if (m_queuedFrames == 0u) {
		m_firstFrameSequence = startSequence;
	}
	const size_t freeFrames = APU_OUTPUT_RING_CAPACITY_FRAMES - m_queuedFrames;
	if (frameCount > freeFrames) {
		const size_t overflowFrames = frameCount - freeFrames;
		m_readFrame += overflowFrames;
		if (m_readFrame >= APU_OUTPUT_RING_CAPACITY_FRAMES) {
			m_readFrame -= APU_OUTPUT_RING_CAPACITY_FRAMES;
		}
		m_queuedFrames -= overflowFrames;
		m_firstFrameSequence += static_cast<i64>(overflowFrames);
	}
	size_t writeFrame = m_readFrame + m_queuedFrames;
	if (writeFrame >= APU_OUTPUT_RING_CAPACITY_FRAMES) {
		writeFrame -= APU_OUTPUT_RING_CAPACITY_FRAMES;
	}
	size_t firstSpan = APU_OUTPUT_RING_CAPACITY_FRAMES - writeFrame;
	if (firstSpan > frameCount) {
		firstSpan = frameCount;
	}
	const size_t firstSamples = firstSpan * 2u;
	std::copy_n(samples, firstSamples, m_queue.data() + writeFrame * 2u);
	const size_t secondSpan = frameCount - firstSpan;
	if (secondSpan > 0u) {
		std::copy_n(samples + firstSamples, secondSpan * 2u, m_queue.data());
	}
	m_queuedFrames += frameCount;
}

u32 ApuOutputRing::readFramePacked() {
	const size_t sampleIndex = m_readFrame * 2u;
	const u32 packed = static_cast<u32>(static_cast<u16>(m_queue[sampleIndex]))
		| (static_cast<u32>(static_cast<u16>(m_queue[sampleIndex + 1u])) << 16u);
	m_readFrame += 1u;
	if (m_readFrame == APU_OUTPUT_RING_CAPACITY_FRAMES) {
		m_readFrame = 0u;
	}
	m_queuedFrames -= 1u;
	m_firstFrameSequence += 1;
	if (m_queuedFrames == 0u) {
		m_readFrame = 0;
	}
	return packed;
}

} // namespace bmsx

File: machine/ts/machine/devices/audio/output_ring.ts
export const APU_OUTPUT_RING_CAPACITY_FRAMES = 3072;
export const APU_OUTPUT_RING_CAPACITY_SAMPLES = APU_OUTPUT_RING_CAPACITY_FRAMES * 2;

export class ApuOutputRing {
	private readonly queue = new Int16Array(APU_OUTPUT_RING_CAPACITY_SAMPLES);
	private readFrame = 0;
	private queuedFramesValue = 0;
	private firstFrameSequenceValue = 0;

	public clear(): void {
		this.readFrame = 0;
		this.queuedFramesValue = 0;
		this.firstFrameSequenceValue = 0;
	}

	public queuedFrames(): number {
		return this.queuedFramesValue;
	}

	public firstFrameSequence(): number {
		return this.firstFrameSequenceValue;
	}

	public write(samples: Int16Array, frameCount: number, startSequence: number): void {
		if (this.queuedFramesValue === 0) {
			this.firstFrameSequenceValue = startSequence;
		}
		const overflowFrames = frameCount - (APU_OUTPUT_RING_CAPACITY_FRAMES - this.queuedFramesValue);
		if (overflowFrames > 0) {
			this.readFrame += overflowFrames;
			if (this.readFrame >= APU_OUTPUT_RING_CAPACITY_FRAMES) {
				this.readFrame -= APU_OUTPUT_RING_CAPACITY_FRAMES;
			}
			this.queuedFramesValue -= overflowFrames;
			this.firstFrameSequenceValue += overflowFrames;
		}
		let writeFrame = this.readFrame + this.queuedFramesValue;
		if (writeFrame >= APU_OUTPUT_RING_CAPACITY_FRAMES) {
			writeFrame -= APU_OUTPUT_RING_CAPACITY_FRAMES;
		}
		let firstSpan = APU_OUTPUT_RING_CAPACITY_FRAMES - writeFrame;
		if (firstSpan > frameCount) {
			firstSpan = frameCount;
		}
		let sourceIndex = 0;
		let targetIndex = writeFrame * 2;
		for (let frame = 0; frame < firstSpan; frame += 1) {
			this.queue[targetIndex] = samples[sourceIndex]!;
			this.queue[targetIndex + 1] = samples[sourceIndex + 1]!;
			targetIndex += 2;
			sourceIndex += 2;
		}
		const secondSpan = frameCount - firstSpan;
		targetIndex = 0;
		for (let frame = 0; frame < secondSpan; frame += 1) {
			this.queue[targetIndex] = samples[sourceIndex]!;
			this.queue[targetIndex + 1] = samples[sourceIndex + 1]!;
			targetIndex += 2;
			sourceIndex += 2;
		}
		this.queuedFramesValue += frameCount;
	}

	public readFramePacked(): number {
		const sampleIndex = this.readFrame * 2;
		const packed = ((this.queue[sampleIndex + 1]! & 0xffff) << 16) | (this.queue[sampleIndex]! & 0xffff);
		this.readFrame += 1;
		if (this.readFrame === APU_OUTPUT_RING_CAPACITY_FRAMES) {
			this.readFrame = 0;
		}
		this.queuedFramesValue -= 1;
		this.firstFrameSequenceValue += 1;
		if (this.queuedFramesValue === 0) {
			this.readFrame = 0;
		}
		return packed >>> 0;
	}
}
